ethcore/src/blockchain/blockchain.rs

```diff
@@ -1299,6 +1389,115 @@ impl BlockChain {
+		let high = Block {
+			hash: high_hash.clone(),
+			// TODO: Should this be an invariant? (i.e. should we raise a programmer error if `get`
+			//       returns `None`)
+			number: self.block_details.read().get(high_hash)?.number,
+		};
+
+		// Have to split this because of lifetimes
+		let best_ancient = self.best_ancient_block.read();
+		let best_ancient = best_ancient.as_ref()?;
+
+		let low = Block {
+			hash: best_ancient.hash,
+			number: best_ancient.number,
+		};
+
+		Some(Gap { low, high })
+	}
+
+	// TODO: have a real error type
+	/// Copies all the data from `other` into `self` and updates the metadata accordingly
+	pub fn migrate_from(&self, other: &Self) -> Result<(), ()> {
+		use std::hash;
+
+		fn migrate_hashmap<K: cmp::Eq + hash::Hash + Clone, V: Clone>(
+			into: &mut HashMap<K, V>,
+			from: &HashMap<K, V>,
+		) {
+			into.extend(from.iter().map(|(k, v)| (k.clone(), v.clone())));
+		}
+
+		fn migrate_kvdb(into: &KeyValueDB, from: &KeyValueDB) -> ::kvdb::Result<()> {
+			const MAX_TRANSACTION_SIZE: usize = 1024;
+
+			let mut transaction = into.transaction();
+			let mut count = 0;
+
+			for &column in [db::COL_HEADERS, db::COL_BODIES].iter() {
+				for (key, value) in from.iter(column) {
+					transaction.put(column, &*key, &*value);
+					count += 1;
+
+					if count == MAX_TRANSACTION_SIZE {
+						count = 0;
+						into.write(transaction)?;
+						transaction = into.transaction();
+					}
+				}
+			}
+
+			into.write(transaction)
+		}
+
+		migrate_hashmap(&mut *self.block_headers.write(), &*other.block_headers.read());
+		migrate_hashmap(&mut *self.block_bodies.write(), &*other.block_bodies.read());
+		migrate_hashmap(&mut *self.block_details.write(), &*other.block_details.read());
+		migrate_hashmap(&mut *self.block_hashes.write(), &*other.block_hashes.read());
+		migrate_hashmap(&mut *self.transaction_addresses.write(), &*other.transaction_addresses.read());
+		migrate_hashmap(&mut *self.block_receipts.write(), &*other.block_receipts.read());
+
+		migrate_kvdb(&*self.db, &*other.db).map_err(|_| ())?;
+
+		let best_block_after_merge = {
+			let (my_best, other_best) = (self.best_block.read(), other.best_block.read());
+
+			if my_best.number > other_best.number {
+				my_best.clone()
+			} else {
+				other_best.clone()
+			}
+		};
+
+		*self.best_block.write() = best_block_after_merge;
+
+		let (my_gap, other_gap) = (
+			self.missing_blocks().ok_or(())?,
+			other.missing_blocks(),
+		);
+
+		let new_gap = other_gap.and_then(|other| my_gap.intersection(&other));
+
+		if let Some(Gap { low, high }) = new_gap {
+			*self.first_block.write() = Some(high.hash);
+			*self.best_ancient_block.write() = Some(
+				BestAncientBlock {
+					hash: low.hash,
+					number: low.number,
+				}
+			);
+
+			let mut transaction = self.db.transaction();
+			transaction.put(db::COL_EXTRA, b"first", &high.hash);
+			transaction.put(db::COL_EXTRA, b"ancient", &low.hash);
+
+			self.db.write(transaction).map_err(|_| ())?;
```

Comment:
There might be an issue here in this specific scenario (`N` is the number of blocks included in a snapshot):

You got a snapshot until block `B1`, and synced ancient blocks from 0 to `AB1`, new block up to `B1 + N`. Some time later, you get a snapshot at `B2 >> B1 + N`. So after the DB migration, you basically have `0` to `AB1`, `B1` to B1 + N`, `B2` to `B2 + N`.

The `new_gap` is then `[AB1 ; B1]`, so `first` would be `B1` which is wrong since you actually miss `[B1 + N ; B2]`.
Even if the gap would be `[AB1 ; B2]`, by writing to DB the blocks `[B1; B1+N]` I think you will run into situations where the old block downloader will stop at `B1` since it hit a known block (cf `insert_unordered_block` method).